pkg/controller/operators/olm/overrides/inject.go

```diff
@@ -0,0 +1,151 @@
+package overrides
+
+import (
+	"errors"
+
+	corev1 "k8s.io/api/core/v1"
+)
+
+// InjectEnvIntoDeployment injects the proxy env variables specified in
+// proxyEnvVar into the container(s) of the given PodSpec.
+//
+// If any Container in PodSpec already defines an env variable of the same name
+// as any of the proxy env variables then it
+func InjectEnvIntoDeployment(podSpec *corev1.PodSpec, envVars []corev1.EnvVar) error {
+	if podSpec == nil {
+		return errors.New("no pod spec provided")
+	}
+
+	for i := range podSpec.Containers {
+		container := &podSpec.Containers[i]
+		container.Env = mergeEnvVars(container.Env, envVars)
+	}
+
+	return nil
+}
+
+func mergeEnvVars(containerEnvVars []corev1.EnvVar, newEnvVars []corev1.EnvVar) (merged []corev1.EnvVar) {
+	merged = containerEnvVars
+
+	for _, newEnvVar := range newEnvVars {
+		existing, found := findEnvVar(containerEnvVars, newEnvVar.Name)
+		if found {
+			existing.Value = newEnvVar.Value
+			continue
+		}
+
+		merged = append(merged, corev1.EnvVar{
+			Name:  newEnvVar.Name,
+			Value: newEnvVar.Value,
+		})
+	}
+
+	return
+}
+
+func findEnvVar(proxyEnvVar []corev1.EnvVar, name string) (foundEnvVar *corev1.EnvVar, found bool) {
+	for i := range proxyEnvVar {
+		if name == proxyEnvVar[i].Name {
+			// Environment variable names are case sensitive.
+			found = true
+			foundEnvVar = &proxyEnvVar[i]
+
+			break
+		}
+	}
+
+	return
+}
+
+// InjectVolumesIntoDeployment injects the provided Volumes
+// into the container(s) of the given PodSpec.
+//
+// If any Container in PodSpec already defines a Volume of the same name
+// as any of the provided Volumes then it will be overwritten.
+func InjectVolumesIntoDeployment(podSpec *corev1.PodSpec, volumes []corev1.Volume) error {
+	if podSpec == nil {
+		return errors.New("no pod spec provided")
+	}
+
+	podSpec.Volumes = mergeVolumes(podSpec.Volumes, volumes)
+
+	return nil
+}
+
+func mergeVolumes(podSpecVolumes []corev1.Volume, newVolumes []corev1.Volume) (merged []corev1.Volume) {
+	merged = podSpecVolumes
+
+	for _, newVolume := range newVolumes {
+		existing, found := findVolume(podSpecVolumes, newVolume.Name)
+		if found {
```

Comment:
I'm wondering with such PodPreset-ish behavior why overwriting was chosen instead of bailing out? Just curious - https://github.com/kubernetes/kubernetes/blob/4d6deea24179a72f36f252664937d2585598ee30/plugin/pkg/admission/podpreset/admission.go#L394.

Reply:
Yeah, maybe it's better to support the feature of overwriting.